easybuild/framework/easyblock.py

```diff
@@ -350,46 +407,22 @@ def fetch_sources(self, sources=None, checksums=None):
         if checksums is None:
             checksums = self.cfg['checksums']
 
-        for index, source in enumerate(sources):
-            extract_cmd, download_filename, source_urls, git_config = None, None, None, None
-
-            if isinstance(source, string_type):
-                filename = source
-
-            elif isinstance(source, dict):
-                # Making a copy to avoid modifying the object with pops
-                source = source.copy()
-                filename = source.pop('filename', None)
-                extract_cmd = source.pop('extract_cmd', None)
-                download_filename = source.pop('download_filename', None)
-                source_urls = source.pop('source_urls', None)
-                git_config = source.pop('git_config', None)
-                if source:
-                    raise EasyBuildError("Found one or more unexpected keys in 'sources' specification: %s", source)
-
-            elif isinstance(source, (list, tuple)) and len(source) == 2:
-                self.log.deprecated("Using a 2-element list/tuple to specify sources is deprecated, "
-                                    "use a dictionary with 'filename', 'extract_cmd' keys instead", '4.0')
-                filename, extract_cmd = source
-            else:
-                raise EasyBuildError("Unexpected source spec, not a string or dict: %s", source)
+        # Single source should be re-wrapped as a list, and checksums with it
+        if isinstance(sources, dict):
+            sources = [sources]
+        if isinstance(checksums, string_type):
+            checksums = [checksums]
 
-            # check if the sources can be located
-            force_download = build_option('force_download') in [FORCE_DOWNLOAD_ALL, FORCE_DOWNLOAD_SOURCES]
-            path = self.obtain_file(filename, download_filename=download_filename, force_download=force_download,
-                                    urls=source_urls, git_config=git_config)
-            if path:
-                self.log.debug('File %s found for source %s' % (path, filename))
-                self.src.append({
-                    'name': filename,
-                    'path': path,
-                    'cmd': extract_cmd,
-                    'checksum': self.get_checksum_for(checksums, filename=filename, index=index),
-                    # always set a finalpath
-                    'finalpath': self.builddir,
-                })
+        # Loop over the list of sources; list of checksums must match >= in size
+        for index, source in enumerate(sources):
+            if source:
```

Comment:
Why the `if source:`? How can this ever be equivalent to `False`?

Reply:
This is a sanity check for a user who puts a dead entry into their `sources` list.  I want to keep the error check here with a clear message, as well as down in `fetch_source()` itself.  The latter is there because it can be called directly, as we do with extensions, so I want an explicit test if someone passes None.

I did rearrange the code slightly so it has the error condition at the top:
```
for index, source in enumerate(sources):
        if source is None:
            raise EasyBuildError("Empty source in sources list at index %d", index)

        got_src = self.fetch_source(source, self.get_checksum_for(checksums=  [...]
```
I tend to prefer this layout when the failure condition is simple, and the success is a big block of code.  Unless you like it for coding style, the "else:" is unnecessary since the error acts like a return/exit.